### ISSUE TITLE
Error 500's and TTS Not Working

### ISSUE THREAD

**Issue by mattnewham** [jishi/node-sonos-http-api#200]
Hey

I think I am doing something wrong here, but I've exhausted my troubleshooting.

I have successfully got setup and running on raspberry pi, I can play/pause content so everything is connected OK, however if I try to send a playlist command, or favourite, node-sonos server says it got an 500. URL is correct checked on multiple. Playlist/Favourite works from app.

Also attempting to play TTS doesn't work either.

Anyone got any ideas what I might be doing wrong?

**Comment by mattnewham** [comment]
I also get "error when fetching favourites" when attempting to play a favourites playlist

**Comment by MrBeanzy** [comment]
@mattnewham Similar for me. For TTS the music is stopped. The MP3 generated (i can download the MP3 successfully through a browser from the node server) but the file is never queued or played. Then the previous music doesn't continue to play.

**Comment by jishi** [comment]
Some logs would be helpful. However, most of my free time is spent on the beta-version that will replace this completely.

**Comment by mattnewham** [comment]
Jimmy

Is the beta available yet? I'd be happy to use that as this doesn't work
for me?

Which logs would you want? The output from the server just states expected
200 but got 500.

**Comment by jishi** [comment]
The beta is available (with the branch named "beta") but the say/sayAll command isn't implemented yet. The presets also doesn't seem to work for people, so it won't work as a daily driver. 

The last 50 rows or so after you have tried to apply the say command might give me a clue on where it breaks.

**Comment by mattnewham** [comment]
Hi Jimmy

The main branch currently isn't working for me on any features, so I'm not
sure I'd lose anything!

The below is from the HTTP server for TTS:

```

Using cached tts message file:
f7ff9e8b7bb2e09b70935a5d785e0cc5d9d0abf0-en-gb.mp3

info: applying preset players=[roomName=Landing, volume=40], state=playing,
playMode=NORMAL, uri=
http://192.168.60.50:5005/tts/f7ff9e8b7bb2e09b70935a5d785e0cc5d9d0abf0-en-gb.mp3

debug: ungrouping Landing to prepare for grouping

STOPPED

STOPPED

STOPPED

STOPPED

STOPPED

STOPPED

debug: emitting group-volume

```

That's all it says. This is When attempting a playlist:

```

Expected statusCode 200, but got 500
```

Is there some additional logging I should be turning on somewhere?

**Comment by jishi** [comment]
Nah, it always log everything so this is it. I'll need to run some tests on my system to see if I can reproduce it. 

But please do check if the playlist and favorite functionality works on the beta. Do note that you need to invoke `npm install` after switching branch since it requires a new version of sonos-discovery.

**Comment by mattnewham** [comment]
Cloned the beta, re ran npm install, server is running. However sending any GET request results in:

/home/pi/.sonoshttp/node-sonos-http-api-beta/node_modules/sonos-discovery/lib/models/Player.js:453
              uri: item.res.$text,
                           ^

TypeError: Cannot read property '$text' of undefined

Feels like I didn't get something right with the sones-discovery package?

**Comment by mattnewham** [comment]
Checked that I am indeed on version 1.0.0 of Sonos Discovery whilst running the beta branch.

Initial startup of the server, log is much more terse than main branch:

> sonos-http-api@1.0.0-beta.1 start /home/pi/.sonoshttp/node-sonos-http-api-beta
> node server.js

sending M-SEARCH...
http server listening on port 5005
no preset file, ignoring...

**Comment by mattnewham** [comment]
More tests:

Sending certain requests works - zones for example works fine. However sending:

>http://192.168.60.50:5005/Landing/favorite/BBC%20Radio%206%20Music

Returns the error above:

>/home/pi/.sonoshttp/node-sonos-http-api-beta/node_modules/sonos->discovery/lib/models/Player.js:453
>uri: item.res.$text,
>^

>TypeError: Cannot read property '$text' of undefined

**Comment by jishi** [comment]
Well, that's a disappointment :) I have tested it at home and it runs, but not on the raspberry though. Back to the drawing board. Odd error, but it is related to parsing of track/source data. Did you have any player that was lacking any source or track? (No music state)?

**Comment by mattnewham** [comment]
The player that I was connecting to at the time had nothing in its playlist, but that might be a normal scenario right? 

I am just attempting to put something in the queue, pause playback and then send the request again.

Stand by!

**Comment by mattnewham** [comment]
OK, so playing music on the Landing speaker from Spotify (or anything). State request returns JSON with correct now playing info. I then send again the above request. Same error

**Comment by mattnewham** [comment]
Actually, different error now:

>/home/pi/.sonoshttp/node-sonos-http-api-beta/lib/actions/say.js:35
 > var state = player.getState();
>                     ^

>TypeError: player.getState is not a function

Not sure why its even calling anything in say.js - I didn't ask it to say anything!

**Comment by jishi** [comment]
BBC radio favorite probably differs a bit in it's metadata, that is an untested case, good catch. 

But they "say" error is weird, was this triggered when you tried to invoke the "favorite" call?

**Comment by mattnewham** [comment]
Yep - same request as above.

I will try with a playlist 1

**Comment by mattnewham** [comment]
Test

Landing player currently has paused song - "State" responds with correct information.

Send the following:

>http://192.168.60.50:5005/Landing/playlist/Hot

Error:

>/home/pi/.sonoshttp/node-sonos-http-api-beta/node_modules/sonos->discovery/lib/models/Player.js:453
 >             uri: item.res.$text,
                           ^

>TypeError: Cannot read property '$text' of undefined

Could these issues be associated to the issues I am experiencing with the main branch not working? Perhaps the speaker is returning a 500 error again with no data, which is causing the response variable to be empty causing the unhanded exception?

**Comment by jishi** [comment]
No, it should not continue if it receives a 500, but when you apply a playlist or a favorite, it first fetches all favorites/playlists and parses them, to find the matching one based on the name. 

If one of the favorite or playlist has a format that I don't support, this could probably happen. I'm guessing that you get the same error if you call `/favorites` or `/playlists` ?

**Comment by mattnewham** [comment]
Yep, same error.

Running a capture on the box to port 1400 shows that the player responds with 200 OK and:

>title&gt;Favorites and >Presets&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;/DIDL>-Lite&gt;</Result><NumberReturned>6</NumberReturned><TotalMatches>6</TotalMatches>><UpdateID>13</UpdateID></u:BrowseResponse></s:Body></s:Envelope>

Snipped that obviously, I've only got TCPDump output so bad formatting, but looks like something is happening correctly at least?

**Comment by mattnewham** [comment]
Cleared out all my playlists but one that I just made (Hot) - same errors. Just looking into the PCAP a bit more to see what is coming back from the player.

**Comment by mattnewham** [comment]
`<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
    <u:BrowseResponse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
        <Result>&lt;DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"&gt;&lt;container id="A:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Attributes&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="S:" parentID="0" restricted="false"&gt;&lt;dc:title&gt;Music Shares&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="Q:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Queues&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="SQ:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Saved Queues&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="R:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Internet Radio&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="FV:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Favorites and Presets&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;/DIDL-Lite&gt;</Result>
        <NumberReturned>6</NumberReturned>
        <TotalMatches>6</TotalMatches>
        <UpdateID>19</UpdateID>
    </u:BrowseResponse>
</s:Body>
</s:Envelope>`

**Comment by mattnewham** [comment]
This looks like the response from the player - it looks like there is an issue with the "sax2.on('tag:container', (item) =>" but I don't know anything about the format of the XML on the players.

```xml
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
    <u:BrowseResponse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
        <Result>&lt;DIDL-Lite xmlns:dc="http://purl.org/dc/elements/1.1/" xmlns:upnp="urn:schemas-upnp-org:metadata-1-0/upnp/" xmlns:r="urn:schemas-rinconnetworks-com:metadata-1-0/" xmlns="urn:schemas-upnp-org:metadata-1-0/DIDL-Lite/"&gt;&lt;container id="A:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Attributes&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="S:" parentID="0" restricted="false"&gt;&lt;dc:title&gt;Music Shares&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="Q:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Queues&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="SQ:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Saved Queues&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="R:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Internet Radio&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;container id="FV:" parentID="0" restricted="true"&gt;&lt;dc:title&gt;Favorites and Presets&lt;/dc:title&gt;&lt;upnp:class&gt;object.container&lt;/upnp:class&gt;&lt;/container&gt;&lt;/DIDL-Lite&gt;</Result>
        <NumberReturned>6</NumberReturned>
        <TotalMatches>6</TotalMatches>
        <UpdateID>19</UpdateID>
    </u:BrowseResponse>
</s:Body>
</s:Envelope>
```

**Comment by jishi** [comment]
Wow, that doesn't seem right at all. I will look into it later, great assistance, thank you very much.

**Comment by mattnewham** [comment]
Some additional observations:

Back to the main branch code - I pulled the XML request/response from the PCAP to see what was causing the error 500 - looks like we're getting UPnP error 701 which is:

>"Transition not available. The immediate transition from current transport state to desired transport state is not supported by this device."

Here are the request/response:

```xml
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
    <u:Browse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
        <ObjectID/>
        <BrowseFlag>BrowseDirectChildren</BrowseFlag>
        <Filter/>
        <StartingIndex/>
        <RequestedCount/>
        <SortCriteria/>
    </u:Browse>
</s:Body>
</s:Envelope>

```

```xml
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
    <s:Fault>
        <faultcode>s:Client</faultcode>
        <faultstring>UPnPError</faultstring>
        <detail>
            <UPnPError xmlns="urn:schemas-upnp-org:control-1-0">
                <errorCode>701</errorCode>
            </UPnPError>
        </detail>
    </s:Fault>
</s:Body>
</s:Envelope>
```

**Comment by jishi** [comment]
Could you also capture the request from the beta branch? For some reason it seems to lack the ObjectID which should be be FV:2 for favorites and SQ: for playlists in the request, which I find very odd (especially for the master branch).

The UPnP error codes are super nonintuitive and are usually very vague in their descriptions :)

**Comment by mattnewham** [comment]
Beta branch request:

```xml
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
    <u:Browse xmlns:u="urn:schemas-upnp-org:service:ContentDirectory:1">
        <ObjectID>0</ObjectID>
        <BrowseFlag>BrowseDirectChildren</BrowseFlag>
        <Filter/>
        <StartingIndex>0</StartingIndex>
        <RequestedCount>0</RequestedCount>
        <SortCriteria/>
    </u:Browse>
</s:Body>
</s:Envelope>
```

Certainly looks to have an object ID, but not the right one?

Maybe I am making the requests wrong? Here is what I am requesting:

>http://192.168.60.50:5005/Landing/playlist/Hot

**Comment by mattnewham** [comment]
This feels like it's going to turn out to be something silly!

**Comment by jishi** [comment]
That ObjectID `0` puzzles me. Can't seem to find anything wrong with the code base, will have to test it out when I get home. 

No, the request looks fine, shouldn't be possible to affect the underlying request in this manner.

**Comment by mattnewham** [comment]
I don't get it either. I've been tracing through the code and you are literally passing the string 'SQ:' into this.browse - I don't understand how it can be missing (on the main branch at least - I don't quite understand the promise stuff on the beta branch, but Im sure the two issues are related)

**Comment by robertw** [comment]
I have it running on a pi no prob as per your error

**Comment by robertw** [comment]
Be sure your updated as pi op system and firmware

**Comment by mattnewham** [comment]
PI is running Jessie lite. Latest and fully updated.

Node is 6.0.0

npm 3.8.6

**Comment by jishi** [comment]
If you could test-run it on node 4 or 5, to rule out any discrepancy in the new node version. Haven't tested 6.0.0 yet.

**Comment by mattnewham** [comment]
What's the best way to run it with an earlier version of node?

**Comment by mattnewham** [comment]
For your ref I think I narrowed it down to the SOAP.Browse.format function - if I print the objectID prior to that it is correct, however in the resultant SOAP the objectID is gone

**Comment by mattnewham** [comment]
TLDR - works on node 5.0.0

Using "n" i switched to node v 5.0.0 and the main branch is working as expected.

**Comment by jishi** [comment]
Hm, I wonder if the RegExp.$1 has been revoked in Node 6, it's a non-standard feature. Ah, so it works in 5, okay, that must be it then. Fairly easy to rewrite without the RegExp.$1 since the matching brackets gets passed as arguments to the replace function. Good catch!

**Comment by mattnewham** [comment]
Hopefully saved you some headaches with your rewrite when people move to node 6!

**Comment by jishi** [comment]
The offending code in node 6 is removed in the beta branch. TTS still not working, but that should be tracked in a separate ticket if necessary.
